tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,354 @@
+                    .format(program_name, container_name, program_status))
+    else:
+        pytest.fail("Failed to find program '{}' in container '{}'"
+                    .format(program_name, container_name))
+
+    logger.info("Waiting until container '{}' is stopped...".format(container_name))
+    pytest_assert(wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+                  CONTAINER_CHECK_INTERVAL_SECS,
+                  check_container_state, duthost, container_name, False),
+                  "Failed to stop container '{}'".format(container_name))
+    logger.info("Container '{}' was stopped".format(container_name))
+
+    logger.info("Waiting until container '{}' is restarted...".format(container_name))
+    restarted = wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
+                           CONTAINER_CHECK_INTERVAL_SECS,
+                           check_container_state, duthost, container_name, True)
+    if not restarted:
+        if is_hiting_start_limit(duthost, container_name):
+            logger.info("{} hits start limit and clear reset-failed flag".format(container_name))
+            duthost.shell("sudo systemctl reset-failed {}.service".format(container_name))
+            duthost.shell("sudo systemctl start {}.service".format(container_name))
+            pytest_assert(wait_until(CONTAINER_RESTART_THRESHOLD_SECS,
+                          CONTAINER_CHECK_INTERVAL_SECS,
+                          check_container_state, duthost, container_name, True),
+                          "Failed to restart container '{}' after reset-failed was cleared".format(container_name))
+        else:
+            pytest.fail("Failed to restart container '{}'".format(container_name))
+
+    logger.info("Container '{}' was restarted".format(container_name))
+
+
+def verify_no_autorestart_with_non_critical_process(duthost, container_name, program_name,
+                                                    program_status, program_pid):
+    """
+    @summary: Kill a non-critical process in a container to verify whether the container
+              remains in the running state
+    """
+    if program_status == "RUNNING":
+        kill_process_by_pid(duthost, container_name, program_name, program_pid)
+    elif program_status in ["EXITED", "STOPPED", "STARTING"]:
+        pytest.fail("Program '{}' in container '{}' is in the {} state, expected 'RUNNING'"
+                    .format(program_name, container_name, program_status))
+    else:
+        pytest.fail("Failed to find program '{}' in container '{}'"
+                    .format(program_name, container_name))
+
+    logger.info("Checking whether container '{}' is still running...".format(container_name))
+    pytest_assert(wait_until(CONTAINER_STOP_THRESHOLD_SECS,
+                  CONTAINER_CHECK_INTERVAL_SECS,
+                  check_container_state, duthost, container_name, True),
+                  "Container '{}' was stopped unexpectedly".format(container_name))
```

Comment:
This check appears incorrect. It is waiting for the container to be running. Instead, it should wait for the container to stop (and since we don't want the container to stop, we *expect* it to time out. Also suggest breaking this into two statements for readability as such:
```
    stopped = wait_until(CONTAINER_STOP_THRESHOLD_SECS,
                         CONTAINER_CHECK_INTERVAL_SECS,
                         check_container_state, duthost, container_name, False)

    pytest_assert(not stopped, "Container '{}' was stopped unexpectedly".format(container_name))
```
Also, the log message above should be reworded:
```
logger.info("Waiting to ensure container '{}' does not stop...".format(container_name))
```

Reply:
At here, after a non-critical process such as `rsyslogd` is killed, the container is expected to still be in the running state. Yes, we don't want the container to stop, we _expect_ it to time out.  Another thing is after `wait_until(...)` timed out, it will return `false` (https://github.com/Azure/sonic-mgmt/blob/9d90cd432641e9acb8de1c71686d1acf9fda1901/tests/common/utilities.py#L52), probably we should use:
`pytest_assert(stopped, "Container '{}' was stopped unexpectedly".format(container_name))`